core/src/main/java/org/radargun/stages/test/SchedulingSelector.java

```diff
@@ -0,0 +1,131 @@
+package org.radargun.stages.test;
+
+import java.lang.reflect.Array;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicIntegerArray;
+import java.util.concurrent.atomic.AtomicLongArray;
+
+import org.radargun.logging.Log;
+import org.radargun.logging.LogFactory;
+import org.radargun.utils.Projections;
+import org.radargun.utils.TimeService;
+
+/**
+ * Based on provided frequency, returns matching invocation from {@link #next()} or blocks the thread calling it.
+ *
+ * @author Radim Vansa &lt;rvansa@redhat.com&gt;
+ */
+public class SchedulingSelector<T> {
+   private final static Log log = LogFactory.getLog(SchedulingSelector.class);
+   private final static boolean trace = log.isTraceEnabled();
+   private final int[] invocations;
+   private final long[] intervals;
+   private final T[] operations;
+   private final int numOperations;
+   private final AtomicLongArray lastIntervals;
+   private final AtomicIntegerArray todoInvocations;
+   private volatile int offset;
+
+   /**
+    * @param operations Returned invocations
+    * @param invocations Number of operations per interval
+    * @param intervals Size of slot, in milliseconds
+    */
+   public SchedulingSelector(T[] operations, int[] invocations, long[] intervals) {
+      if (operations.length != invocations.length) throw new IllegalArgumentException();
+      this.operations = operations;
+      this.invocations = invocations;
+      this.intervals = intervals;
+      numOperations = operations.length;
+      lastIntervals = new AtomicLongArray(numOperations);
+      todoInvocations = new AtomicIntegerArray(numOperations);
+   }
+
+   /**
+    * @return
+    * @throws InterruptedException
+    */
+   public T next() throws InterruptedException {
+      WAIT_LOOP: for (;;) {
```

Comment:
It is desired to have the thread waiting here until it gets the operation - this is the part that should keep threads blocked. It is perfectly valid to have the thread waiting here for the whole duration of test. The only undesired property is that the threads have to wake up every 1 ms too see that the time has advanced and there may be more work to do - this is a place for optimization.
If we want to forcefully stop the thread, we interrupt it - and I can guarantee that the thread reacts to interruption.